polygeist/tools/cgeist/driver.cc

```diff
@@ -373,7 +363,7 @@ static int optimize(mlir::MLIRContext &context,
   if (DetectReduction)
     optPM.addPass(polygeist::detectReductionPass());
 
-  if (!Opt0) {
+  if (!OptimizationLevel || *OptimizationLevel != llvm::OptimizationLevel::O0) {
```

Comment:
Is `!OptimizationLevel` intended here? This would cause an access to an optional without value in the second part of the condition, wouldn't it?

Reply:
Seems intended. When optional without value, `!OptimizationLevel` is true, `*OptimizationLevel != llvm::OptimizationLevel::O0` is not evaluated, and the body is entered.

Reply:
That means we perform the optimizations if no optimization level was given by the user. Is this the intended default? And which level do we optimize in that case, `-O2`?

Reply:
I think we should run `-O2` in that case, although we're not running anything currently. I did not want to change behavior, but I think I'll introduce that change.

Reply:
I agree when no `-O` option is given, then optimization level should be `-O0`. When `-O` is given, then optimization level should be `-O1`.

Reply:
- Default: `-O0` (with `-fsycl`, we will always be passed `-O2` by default)

To avoid updating almost every non-SYCL test, we add a temporary `-nopt-mlir` flag to be removed when tests are updated.

Reply:
> Default: -O0 (with -fsycl, we will always be passed -O2 by default)

I tried to compile a SYCL file with `-fsycl` and `-###` using `clang`, I see that there are no optimization options passed to `cgeist` explicitly. What do you mean by `we will always be passed -O2 by default`?

> To avoid updating almost every non-SYCL test...

Why do they need to be updated? Their original behavior is not `-O0`?

Reply:
> I tried to compile a SYCL file with `-fsycl` and `-###` using `clang`, I see that there are no optimization options passed to `cgeist` explicitly. What do you mean by `we will always be passed -O2 by default`?

Interesting, I swore that was the case. No problem, though, as correct behavior is already codified here.



> Why do they need to be updated? Their original behavior is not `-O0`?

It isn't, as MLIR simplification happens always unless `-O0` is explicitly stated.